pkg/meta/sql_bak.go

```diff
@@ -0,0 +1,1050 @@
+	return map[string]*int64{
+		usedSpace:     &c.UsedSpace,
+		totalInodes:   &c.UsedInodes,
+		"nextInode":   &c.NextInode,
+		"nextChunk":   &c.NextChunk,
+		"nextSession": &c.NextSession,
+		"nextTrash":   &c.NextTrash,
+	}
+}
+
+type sqlCounterDS struct {
+	dumpedSeg
+}
+
+func (s *sqlCounterDS) dump(ctx Context, ch chan *dumpedResult) error {
+	meta := s.meta.(*dbMeta)
+	var rows []counter
+	if err := meta.execStmt(ctx, s.txn, func(s *xorm.Session) error {
+		rows = rows[:0]
+		return s.Find(&rows)
+	}); err != nil {
+		return err
+	}
+	counters := &pb.Counters{}
+	fieldMap := getSQLCounterFields(counters)
+	for _, row := range rows {
+		if fieldPtr, ok := fieldMap[row.Name]; ok {
+			*fieldPtr = row.Value
+		}
+	}
+	if err := dumpResult(ctx, ch, &dumpedResult{s, counters}); err != nil {
+		return err
+	}
+	logger.Debugf("dump %s result %+v", s, counters)
+	return nil
+}
+
+type sqlSustainedDS struct {
+	dumpedSeg
+}
+
+func (s *sqlSustainedDS) dump(ctx Context, ch chan *dumpedResult) error {
+	var rows []sustained
+	if err := s.meta.(*dbMeta).execStmt(ctx, s.txn, func(s *xorm.Session) error {
+		rows = rows[:0]
+		return s.Find(&rows)
+	}); err != nil {
+		return err
+	}
+	ss := make(map[uint64][]uint64)
+	for _, row := range rows {
+		ss[row.Sid] = append(ss[row.Sid], uint64(row.Inode))
+	}
+
+	pss := &pb.SustainedList{
+		List: make([]*pb.Sustained, 0, len(ss)),
+	}
+	for k, v := range ss {
+		pss.List = append(pss.List, &pb.Sustained{Sid: k, Inodes: v})
+	}
+
+	if err := dumpResult(ctx, ch, &dumpedResult{s, pss}); err != nil {
+		return err
+	}
+	logger.Debugf("dump %s num %d", s, len(ss))
+	return nil
+}
+
+type sqlDelFileDS struct {
+	dumpedSeg
+}
+
+func (s *sqlDelFileDS) dump(ctx Context, ch chan *dumpedResult) error {
+	var rows []delfile
+	if err := s.meta.(*dbMeta).execStmt(ctx, s.txn, func(s *xorm.Session) error {
+		rows = rows[:0]
+		return s.Find(&rows)
+	}); err != nil {
+		return err
+	}
+	delFiles := &pb.DelFileList{List: make([]*pb.DelFile, 0, len(rows))}
+	for _, row := range rows {
+		delFiles.List = append(delFiles.List, &pb.DelFile{Inode: uint64(row.Inode), Length: row.Length, Expire: row.Expire})
+	}
+	if err := dumpResult(ctx, ch, &dumpedResult{s, delFiles}); err != nil {
+		return err
+	}
+	logger.Debugf("dump %s num %d", s, len(delFiles.List))
+	return nil
+}
+
+type sqlSliceRefDS struct {
+	dumpedSeg
+	pools []*sync.Pool
+}
+
+func (s *sqlSliceRefDS) dump(ctx Context, ch chan *dumpedResult) error {
+	eg, _ := errgroup.WithContext(ctx)
+	eg.SetLimit(s.opt.CoNum)
+
+	taskFinished := false
+	psrs := &pb.SliceRefList{List: make([]*pb.SliceRef, 0, 1024)}
+	for start := 0; !taskFinished; start += sqlBatchSize {
+		nStart := start
+		eg.Go(func() error {
+			var rows []sliceRef
+			if err := s.meta.(*dbMeta).execStmt(ctx, s.txn, func(s *xorm.Session) error {
+				rows = rows[:0]
+				return s.Where("refs != 1").Limit(sqlBatchSize, nStart).Find(&rows) // skip default refs
```

Comment:
`refs > 1`?

Reply:
ref == 0 means that the part to be reclaimed is also needed.